File: problems/3sum/solution.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>>res;
        if(n<3)return res;
        sort(nums.begin(),nums.end());
        for(int itr=0 ; itr<n; itr++){
            // Skip duplicates
            if(itr>0 && nums[itr]==nums[itr-1])continue;
            int leftPtr=itr+1 , rightPtr = n-1;
            while(leftPtr<rightPtr){
                int sum = nums[itr]+nums[leftPtr]+nums[rightPtr];
                if(sum==0){
                    res.push_back({nums[itr],nums[leftPtr],nums[rightPtr]});
                    leftPtr++;
                    rightPtr--;
                    //Skip duplicates
                    while(leftPtr<rightPtr && nums[leftPtr]==nums[leftPtr-1])leftPtr++;
                    while(leftPtr<rightPtr && nums[rightPtr]==nums[rightPtr+1])rightPtr--;
                }
                else if(sum<0)leftPtr++;
                else rightPtr--;
            }
        }
        return res;
        
    }
};

File: problems/3sum/solution.ts
function threeSum(nums: number[]): number[][] {
    const n = nums.length;
    if(n<3)return [];
    let res:number[][]=[];
    nums.sort((a,b)=>a-b);
    for(let i:number=0;i<n;i++){
        if(i>0 && nums[i]==nums[i-1])continue;
        let left= i+1, right = n-1;
        while(left<right){
            let sum : number = nums[i]+nums[left]+nums[right];
            if(sum==0){
                res.push([nums[i],nums[left],nums[right]]);
                left++;
                right--;
                while(left<right && nums[left]==nums[left-1])left++;
                while(left<right && nums[right]==nums[right+1])right--;
            } else if(sum<0)left++;
            else right--;
        }
    }
    return res;
};